File: Arrays/sliding-window-and-two-pointers/1423-maximum-points.py
def maximumPoints(nums, k):
    leftSum = 0
    rightSum = 0
    maxSum = 0
    rightIndex = len(nums) - 1
    
    # Step 1: Take first k elements from left
    for i in range(k):
        leftSum += nums[i]
        
    maxSum = leftSum

    # Step 2: Slide one by one from right
    for i in range(k-1, -1, -1):
        leftSum -= nums[i]
        rightSum += nums[rightIndex]
        maxSum = max(maxSum, leftSum + rightSum)
        rightIndex -= 1
        
    return maxSum


def main():
    nums = [12, 1, 4, 2, 9, 3, 9, 1, 7]
    k = 4
    ans = maximumPoints(nums, k)
    print(ans)   #  Output: 29


main()

File: Arrays/sliding-window-and-two-pointers/1423-maximum-points.js
const maximumPoints = (nums, k, n) => {
    let leftSum = 0;
    let rightSum = 0;
    let maxSum = 0;
    let rightIndex = n-1;

    for(let i = 0; i < k; i++) {
        leftSum += nums[i];
    }

    maxSum = leftSum;

    for(let i = k-1; i>= 0; i--) {
        leftSum -= nums[i];
        rightSum += nums[rightIndex];
        maxSum = Math.max(maxSum, leftSum + rightSum);
        rightIndex--;
    }
    return maxSum;
}

function main() {
    const nums = [1,2,33,1,3,7,2,3,90];
    const n = nums.length;
    const k = 3;
    const ans = maximumPoints(nums, k, n);
    console.log(ans); // 95
}

main();